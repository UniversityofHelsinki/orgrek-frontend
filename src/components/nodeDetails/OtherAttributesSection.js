import React from 'react';
import EditableAccordion from '../EditableAccordion';
import { useTranslation } from 'react-i18next';
import useAttributes from '../../hooks/useAttributes';
import AttributesTable from '../attributes/AttributesTable';
import Placeholder from '../Placeholder';
import { authActions } from '../../auth';
import EditableContent from '../EditableContent';
import OtherAttributesEditor from './OtherAttributesEditor';
import { toFormValues } from '../../utils/attributeUtils';
import { useOtherAttributes } from '../../hooks/useOtherAttributes';
import useFilterAttributesByDate from '../../hooks/useFilterAttributesByDate';

const OtherAttributesSection = () => {
  const { t } = useTranslation();
  const { nodeOtherAttributes, isFetching } = useOtherAttributes();
  const initialValues = toFormValues(nodeOtherAttributes);

  // In view mode filter history and future depending on selection
  const sortedAndFilteredData = useFilterAttributesByDate(nodeOtherAttributes);

  const title = t('other_attributes');
  const empty = nodeOtherAttributes.length === 0;

  if (isFetching) {
    return <EditableAccordion title={title} loading />;
  }

  const handleSubmit = (values) => {
    return null;
  };
  return (
    <EditableAccordion
      title={title}
      loading={isFetching}
      defaultExpanded={!empty}
    >
      <EditableContent
        editorComponent={<OtherAttributesEditor />}
        initialValues={initialValues}
        onSubmit={handleSubmit}
        successMessage={t('typeInfo.saveSuccess')}
        errorMessage={t('typeInfo.saveError')}
        authActions={authActions.otherAttributes}
      >
        <Placeholder
          empty={empty}
          placeholder={t('nodeDetailsSection.noAttributes')}
        >
          <AttributesTable data={sortedAndFilteredData} summary={title} />
        </Placeholder>
      </EditableContent>
    </EditableAccordion>
  );
};

export default OtherAttributesSection;
