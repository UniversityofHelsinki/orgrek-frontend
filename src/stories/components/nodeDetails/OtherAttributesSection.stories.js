import {
  mockGetAttributeKeys,
  mockGetOtherAttributes,
  mockSaveOtherAttributes,
  withMockStore,
} from '../../../mockStore';
import OtherAttributesSection from '../../../components/nodeDetails/OtherAttributesSection';
import { waitFor, within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';

// Use a fixed date to ensure that tests always have a consistent result
const now = new Date('2023-03-22T14:28:00+0200');

const nodeId = '12345678';
const selectedHierarchy = 'virallinen';
const keys = ['publicity', 'attribute2'];

export default {
  component: OtherAttributesSection,
  parameters: {
    systemTime: now,
    reactRouter: {
      searchParams: {
        uid: nodeId,
      },
    },
  },
};

export const Default = {
  parameters: {
    msw: {
      handlers: [
        mockGetOtherAttributes(nodeId, selectedHierarchy, [
          {
            id: 1001,
            nodeId: '1',
            key: 'publicity',
            value: 'julkinen',
            startDate: null,
            endDate: null,
          },
          {
            id: 1002,
            nodeId: '1',
            key: 'attribute2',
            value: 'previous value',
            startDate: '2013-01-01',
            endDate: '2015-06-30',
          },
          {
            id: 1002,
            nodeId: '1',
            key: 'attribute2',
            value: 'current value',
            startDate: '2015-07-01',
            endDate: '2023-12-31',
          },
          {
            id: 1002,
            nodeId: '1',
            key: 'attribute2',
            value: 'next value',
            startDate: '2024-01-01',
            endDate: null,
          },
        ]),
        mockSaveOtherAttributes(nodeId),
        mockGetAttributeKeys(
          {
            selectedHierarchies: selectedHierarchy,
            sections: ['other_attributes'],
          },
          keys
        ),
      ],
    },
  },
  decorators: [
    withMockStore({
      tree: {
        selectedHierarchy,
      },
    }),
  ],
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await waitFor(async () => {
      await expect(canvas.getByText('Muut attribuutit')).toBeInTheDocument();
    });
    await expect(canvas.queryByText('Julkinen')).toBeInTheDocument();
    await expect(canvas.queryByText('current value')).toBeInTheDocument();
    await expect(canvas.queryByText('previous value')).not.toBeInTheDocument();
    await expect(canvas.queryByText('next value')).not.toBeInTheDocument();
  },
};

export const Empty = {
  ...Default,
  parameters: {
    msw: {
      handlers: [
        mockGetOtherAttributes(nodeId, selectedHierarchy, []),
        mockSaveOtherAttributes(nodeId),
        mockGetAttributeKeys(
          {
            selectedHierarchies: selectedHierarchy,
            sections: ['other_attributes'],
          },
          keys
        ),
      ],
    },
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await waitFor(async () => {
      await expect(
        canvas.getByText('Ei muita attribuutteja')
      ).toBeInTheDocument();
    });
  },
};

export const ShowHistory = {
  ...Default,
  decorators: [
    withMockStore({
      nvrd: {
        showHistory: true,
      },
      tree: {
        selectedHierarchy,
      },
    }),
  ],
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await waitFor(async () => {
      await expect(canvas.getByText('previous value')).toBeInTheDocument();
    });
  },
};

export const HistoryAndFuture = {
  ...Default,
  decorators: [
    withMockStore({
      nvrd: {
        showHistory: true,
        showComing: true,
      },
      tree: {
        selectedHierarchy,
      },
    }),
  ],
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await waitFor(async () => {
      await expect(canvas.getByText('previous value')).toBeInTheDocument();
    });
    await expect(canvas.getByText('current value')).toBeInTheDocument();
    await expect(canvas.getByText('next value')).toBeInTheDocument();
  },
};
