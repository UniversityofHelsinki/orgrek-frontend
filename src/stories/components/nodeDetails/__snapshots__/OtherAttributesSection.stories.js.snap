// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/nodeDetails/OtherAttributesSection Default smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "expanded": true,
      "name": "Muut attribuutit",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Attribuutti",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Arvo",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Voimassaolo",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Julkisuus",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Julkinen",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Voimassaoloa ei määritetty",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Additional attribute",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "some value",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "1.1.2013 - 30.6.2015",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-nodedetails-otherattributessection--default",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/nodeDetails/OtherAttributesSection Empty smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "expanded": true,
      "name": "Muut attribuutit",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Ei attribuutteja",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-nodedetails-otherattributessection--empty",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/nodeDetails/OtherAttributesSection HistoryAndFuture smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "expanded": true,
      "name": "Muut attribuutit",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Attribuutti",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Arvo",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Voimassaolo",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Julkisuus",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Julkinen",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Voimassaoloa ei määritetty",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Additional attribute",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "previous value",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "1.1.2013 - 30.6.2015",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Additional attribute",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "current value",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "1.7.2015 - 31.12.2023",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Additional attribute",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "next value",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "1.1.2024 alkaen",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-nodedetails-otherattributessection--history-and-future",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/nodeDetails/OtherAttributesSection ShowHistory smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "expanded": true,
      "name": "Muut attribuutit",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Attribuutti",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Arvo",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Voimassaolo",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Julkisuus",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Julkinen",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Voimassaoloa ei määritetty",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Additional attribute",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "previous value",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "1.1.2013 - 30.6.2015",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Additional attribute",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "current value",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "1.7.2015 - 31.12.2023",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-nodedetails-otherattributessection--show-history",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;
